### ISSUE TITLE
Fine-tuning to predict emotion labels

### ISSUE THREAD

**Issue by tiimoS** [bfelbo/DeepMoji#54]
Hi, 
I have been trying to use DeepMoji to predict emotion labels for a given text. I have 7 labels I would like to classify: 
```python
{'0': 'neutral', '1': 'angry', '2': 'disgusted', '3': 'afraid', '4': 'joyful', '5': 'sad', '6': 'surprised'}
```

For that purpose, I have used the principles from finetune_youtube_last.py to fine-tune the model for this task with nb_classes=7 and the accuracy measure. The fine-tuning worked nicely and it reported an accuracy of about 85% in the end. The training data was split 0.7, 0.1, 0.2 for training, test and validation (default params). 

I have used the DailyDialogues dataset with the following emotion distribution:
```python
{ "neutral": 72143, "joyful": 11182, "surprised": 1600, "afraid": 146, "disgusted": 303, "sad": 969, "angry": 827}
```
I am fully aware that the unbalanced distribution of the emotions might yield in poorer results. 

Now, I have built the classifier using the deepmoji_transfer model passing it the nb_classes=7 and the path to my fine-tuned model.  I am trying to predict the emotion label similar to what is done in [examples/score_texts_emojis.py](https://github.com/bfelbo/DeepMoji/blob/master/examples/score_texts_emojis.py)

```python
tokenized, _, _ = st.tokenize_sentences(TEST_SENTENCES)
model = deepmoji_transfer(maxlen, PRETRAINED_PATH, nb_classes=7)
prob = model.predict(tokenized)[0]
```

Given the message : 'I'm sorry I'm so late . I had a really bad day .' with the true label '5: sad', the model predicts the following prob: 
```python
prob = [ 0.09569884  0.0398518  -0.04161     0.00866034  0.06039366 -0.07945964
 -0.07649311]
```

Yielding in the following sorted predictions (using top_elements(prob, 7)): 
```python
sorted_idx = [0 4 1 3 2 6 5]
sorted_labels = ['neutral', 'joyful', 'angry', 'afraid', 'disgusted', 'surprised', 'sad']
```

So the predicted label would be 'neutral' here.

I have several questions/problems with this classifier that you might help me with: 
- What do the negative probabilities in the prob array mean? e.g. the prob -0.04161 of the label  '2': 'disgusted'. 
- Do you see any issues in my procedure for predicting the label from the code above? 
- I observe non-deterministic behaviour. Every time I reload the model, it yields in different predictions. For example when running the prediction several times for the sentence above whilst re-loading the model from scratch each time, it yields in a different prediction almost every time. If you run a prediction several times within the same model, it always produces the same predictions (OK). 
- When I validate the model on my test set, I get a precision of about 13.4% compared to the accuracy of 85% during fine-tuning. How can this be? I don't think I'm overfitting the training data since when I run my validation on the training data I get a precision of just about 2.6%

Your support is highly appreciated! Thanks in advance for the efforts.

**Comment by bfelbo** [comment]
The probability output you listed don't make sense. Probabilities can never be negative. It'd look into that first before digging into anything else :)

**Comment by bfelbo** [comment]
I'm assuming you're supplying our own `PRETRAINED_PATH` to `deepmoji_transfer()` and not the default DeepMoji one?

**Comment by bfelbo** [comment]
This seems to be more of a usage question than a problem with the package so I'm closing it here.  However, I'm happy to help you brainstorm what the issue could be if you provide more info.

**Comment by tiimoS** [comment]
@bfelbo thanks for the feedback! 
I have setup everything from scratch again and now the probabilities are all positive :)

> I'm assuming you're supplying our own `PRETRAINED_PATH` to `deepmoji_transfer()` and not the default DeepMoji one?
Yes, I have used finetune_dataset.py to train the model over a conversation dataset with 6 emotion labels. The finetuning worked alright and I got reported a test accuracy of about 70% at the end of the fine-tuning, which would be okay!

However, I observe strange behaviour when running the predictions: 

1. When I run my predictions on the same test that was used during fine-tuning, I only get accuracies of around 10%.  I use nearly the same code as in score_emojis.py, the only thing different is the usage of deepmoji_transfer instead of deepmoji_emojis and the PRETRAINED_PATH points to my fine-tuned .hdf5 weights file. 
```python

class Predictor:
    vocabulary = set_vocabulary()

    def __init__(self):
        self.model = deepmoji_transfer(nb_classes=6, maxlen=30, weight_path=FINETUNED_MODEL_PATH, extend_embedding=1058)
        self.tokenizer = SentenceTokenizer(Predictor.vocabulary, Predictor.max_len)
        self.emotions_dict = {0: 'angry', 1: 'disgusted', 2: 'afraid', 3: 'joyful', 4: 'sad', 5: 'surprised'}

    def predict_emotions(self, message):
        message = unicode(message)
        tokenized, _, _ = self.tokenizer.tokenize_sentences([message])
        prob = self.model.predict(tokenized)[0]
        top_prediction_idx_sorted = top_elements(prob, 6)
        prediction_labels_sorted = [self.emotions_dict[idx] for idx in top_prediction_idx_sorted]
        probabilities_sorted = [prob[idx] for idx in top_prediction_idx_sorted]

        return prediction_labels_sorted, probabilities_sorted


def top_elements(array, k):
    ind = np.argpartition(array, -k)[-k:]
    return ind[np.argsort(array[ind])][::-1]
```

2. The predictions produced by prob = self.model.predict(tokenized)[0] are not deterministic. When I initialize the model multiple times, I get different predictions on the same sentence with each model. I don't think this behaviour is expected? 

Any support is highly appreciated !

**Comment by bfelbo** [comment]
```
The predictions produced by prob = self.model.predict(tokenized)[0] are not deterministic. When I initialize the model multiple times, I get different predictions on the same sentence with each model. I don't think this behaviour is expected?
```

This sounds strange and definitely not intended. My guess would be that the model is not loading all your weights correctly and some weights in the model are getting randomly initialized, leading to this non-deterministic behavior. Make sure you check that the weights for your final softmax layer are being loaded correctly :)

**Comment by tiimoS** [comment]
> Make sure you check that the weights for your final softmax layer are being loaded correctly :)

I assume this would probably be the code snippet below from the model_def.py file. I see that the softmax layer is being excluded when you pass your own weights. I have removed this and now the model is deterministic! Thanks a lot for the help, highly appreciated!

```python
def deepmoji_transfer(nb_classes, maxlen, weight_path=None, extend_embedding=0,
                      embed_dropout_rate=0.25, final_dropout_rate=0.5,
                      embed_l2=1E-6):
    model = deepmoji_architecture(nb_classes=nb_classes,
                                  nb_tokens=NB_TOKENS + extend_embedding,
                                  maxlen=maxlen, embed_dropout_rate=embed_dropout_rate,
                                  final_dropout_rate=final_dropout_rate, embed_l2=embed_l2)

    if weight_path is not None:
        load_specific_weights(model, weight_path,
                              exclude_names=['softmax'],
                              extend_embedding=extend_embedding)
    return model
```

**Comment by bfelbo** [comment]
It's assumed that the `deepmoji_transfer` method is called before doing the transfer learning (i.e. training a new softmax layer). See this description at the top of the method:
```
Loads the pretrained DeepMoji model for finetuning/transfer learning. Does not load weights for the softmax layer.
```
If you're loading all the weights, you could use `deepmoji_architecture()` directly to define the architecture. You can then call `load_specific_weights()` with `exclude_names=[]` or use Keras' `model.load_weights()` to load all weights.

**Comment by tiimoS** [comment]
Thanks a lot for the support! I am now using `deepmoji_architecture()` and `load_specific_weights()` as you proposed. Now everything works smoothly.
